pkg/apis/policy/v1alpha1/upstreamtraffic.go

```diff
@@ -0,0 +1,103 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// UpstreamTraffic defines the policy applicable to traffic destined
+// to an upstream host.
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type UpstreamTraffic struct {
+	// Object's type metadata
+	metav1.TypeMeta `json:",inline"`
+
+	// Object's metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the UpstreamTraffic policy specification
+	// +optional
+	Spec UpstreamTrafficSpec `json:"spec,omitempty"`
+}
+
+// UpstreamTrafficSpec defines the upstream traffic policy specification.
+type UpstreamTrafficSpec struct {
+	// Host the upstream traffic is directed to.
+	// Must either be an FQDN corresponding to the upstream service
+	// or the name of the upstream service. If only the service name
+	// is specified, the FQDN is derived from the service name and
+	// the namespace of the UpstreamTraffic rule.
+	Host string `json:"host"`
+
+	// ConnectionSettings specifies the connection settings for traffic
+	// directed to the upstream host.
+	// +optional
+	ConnectionSettings ConnectionSettingsSpec `json:"connectionSettings,omitempty"`
+}
+
+// ConnectionSettingsSpec defines the connection settings for an
+// upstream host.
+type ConnectionSettingsSpec struct {
+	// TCP specifies the TCP level connection settings.
+	// Applies to both TCP and HTTP connections.
+	// +optional
+	TCP *TCPConnectionSettings `json:"tcp,omitempty"`
+
+	// HTTP specifies the HTTP level connection settings.
+	// +optional
+	HTTP *HTTPConnectionSettings `json:"http,omitempty"`
+}
+
+// TCPConnectionSettings defines the TCP connection settings for an
+// upstream host.
+type TCPConnectionSettings struct {
+	// MaxConnections specifies the maximum number of TCP connections
+	// allowed to the upstream host.
+	// Defaults to 1024 if not specified.
```

Comment:
I wonder if it's worth distinguishing whether it's OSM or Envoy who's actually applying this default. I could interpret this as "if I don't specify this when the resource is created and I read it later, it will have the default" (where OSM is expected to apply the default), but if we simply pass the value to Envoy and let it apply the default then that wouldn't be the case.

Reply:
This is an OSM API, not Envoy's. So the defaults will be applied by OSM controller regardless of what Envoy uses as defaults. This is primarily because some Envoy defaults may not be suitable defaults. At present, I have set these to match Envoy defaults, but these could change later, e.g. to 2^32 - 1 based on the most common use case. At the moment it's hard to reason how sensible these defaults are. Given the fields are optional in Envoy with defaults values if unspecified, I prefer to explicitly specify defaults from within OSM for such fields.